dir/base.go

```diff
@@ -0,0 +1,102 @@
+package dir
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+)
+
+const (
+	notation = "notation"
+)
+
+var (
+	// for mocking
+	goos          = runtime.GOOS
+	userConfigDir = os.UserConfigDir
+	userCacheDir  = os.UserCacheDir
+	getenv        = os.Getenv
+
+	// systemLibexec directory for binaries not meant to be executed directly by users' shell or scripts
+	systemLibexec string
+	// systemConfig directory for configurations
+	systemConfig string
+
+	// userLibexec is user level libexec directory
+	userLibexec string
+	// userConfig is user level config directory
+	userConfig string
+	// userCache for user-specific cache
+	userCache string
+
+	configDir Config
+	cacheDir  Cache
+)
+
+// LoadPath function defines the directory for NotationLibexec, NotationConfig, NotationCache
+//
+// User MUST call this function before access any path.
+func LoadPath() {
+	var err error
+	// set system config and libexec
+	switch goos {
+	case "darwin":
+		systemConfig = "/Library/Application Support/notation"
+		systemLibexec = "/usr/local/lib/notation"
+	case "windows":
+		systemConfig = getenv("ProgramData")
+		if systemConfig == "" {
+			// unsupported OS
+			panic("environment variable `ProgramData` is not set.")
+		}
+		systemConfig = filepath.Join(systemConfig, notation)
+
+		systemLibexec = getenv("ProgramFiles")
+		if systemLibexec == "" {
+			// unsupported OS
+			panic("environment variable `ProgramFiles` is not set.")
+		}
+		systemLibexec = filepath.Join(systemLibexec, notation)
+
+	default:
+		systemConfig = "/etc/notation"
+		systemLibexec = "/usr/libexec/notation"
+	}
+
+	// set user config
+	userConfig, err = userConfigDir()
+	if err != nil {
+		panic(err)
+	}
+	userConfig = filepath.Join(userConfig, notation)
+	// set user libexec
+	userLibexec = userConfig
+	// set user cache
+	userCache, err = userCacheDir()
+	if err != nil {
+		panic(err)
+	}
+	userCache = filepath.Join(userCache, notation)
+
+	// set unionDirFS
+	// TODO(JeyJeyGao): The user/system directory priority may change later (https://github.com/notaryproject/notation/issues/203)
+	configDir = NewConfig(NewUnionDirFS(userConfig, systemConfig))
+	cacheDir = NewCache(NewUnionDirFS(userCache))
+}
+
+// ConfigDir returns Config struct for accessing path in config directory
+func ConfigDir() Config {
+	return configDir
+}
+
+// CacheDir returns the Cache struct for accessing path in cache directory
+func CacheDir() Cache {
+	return cacheDir
+}
+
+// PluginFS returns the UnionDirFS for notation plugins
+func PluginFS(dirs ...string) UnionDirFS {
+	dirs = append(dirs, filepath.Join(userLibexec, "plugins"))
+	dirs = append(dirs, filepath.Join(systemLibexec, "plugins"))
+	return NewUnionDirFS(dirs...)
+}
```

Comment:
Can we expose a unified interface all these paths?

Reply:
Done.